Start turning health-care application into XML

awong-dev opened the pull request.

awong-dev committed:
WIPR: Start turning form into xml.

awong-dev committed:
Start of auto-save.

awong-dev committed:
Basic xml test fixture work going.

awong-dev committed:
Move to Mocha/Chai.

awong-dev committed:
Stub out the XML equivalence tests.

Now this is ready for transitioning to a real test.
In other news, working with XML in javascript is hard.

awong-dev committed:
Load healthcare form as a fixture.

awong-dev committed:
Hack loading test page via an iframe.

Since this javascript is mostly a test of of snagging things from a
form it makes no sense to create a dummy fixture that looks like the
main page.

Instead, load the actual page in an iframe and inspect the contentWindow
relying on the fact that everything is on the same origin.

Revert the testing framework to Jasmine because mocha async handling
has a deadlock with PhantomJS when waiting for the onload event of an
iframe. To be fair, Jasmine also looks to have a deadlock, but on the
timeout, somehow Jasmine actually manages to load the iframe whereas
in mocha seems to get stuck.

Lastly, write a custom XML matches for Jasmine. For some reason, no one
in the Javascript ecosystem (either in Jasmine or Mocha) has written a
sane matcher.

akainic commented on _health-care/_js/_form.js:
```diff
@@ -0,0 +1,348 @@
+var HealthApp = (function() {
+  "use strict";
+
+  var identityTransform = function(formData, arg) { return arg; }
```
Might want to add a `;` at the end, though it probably doesn't matter because I think it gets automatically added anyway.

awong-dev replied:
Good catch. We should add a JS linter....

akainic commented on _health-care/_js/_form.js:
```diff
@@ -0,0 +1,348 @@
+    }
+
+    return xmlDoc;
+  }
```
`;`

awong-dev replied:
grrr...javascript. :(

akainic commented on _health-care/_js/_form.js:
```diff
@@ -0,0 +1,348 @@
+    getFormRoot: getFormRoot,
+    saveForm: saveForm,
+    build1010ezXml: build1010ezXml,
+    initForm: initForm,
```
Don't need that last `,`

awong-dev replied:
Fixed.

akainic commented:
@awong-dev Honestly this looks really great! Don't really have too many comments, in fact I learned a few things from looking through this! I'm not as familiar with Jasmine and Karma so I don't feel confident reviewing that, but the logic makes sense to me. Really awesome :)

awong-dev committed:
Switch back to mocha/chai.

It wasn't a deadlock between Mocha and PhantomJS. Mocha just had too
short of a timeout. For some reason, PhantomJS takes 4 seconds on this
machine to load the iframe from the karma server. It's unclear why this
is happening, but for now, moving back to Mocha.

Also fix broken closing tag. Hah HTMLProof.

awong-dev commented:
@akainic Thanks for the review! I addressed your comments and also moved us back to mocha/chai.

Can you give a final review and then, assuming no problems, stamp it with an LGTM and merge? :dancer:

awong-dev committed:
Extend the timeout to 10s cause Travis is slow.

awong-dev committed:
Use a CSP to disable external resource loads.

PhantomJS's slowdown is caused by external resource loads to things like
Google analytics and usasearch which are embedded in the headers and
footers of this site. It causes a runtime of > 4s per test due to lack of
cacheing. It also breaks hermeticity of the test. These modules should
not be required for most of our tests.

Using a CSP header that enforces a same-origin only policy for *all* resource
loads provides a way to speed up tests w/o source code changes. It means
tests for things depend on external resources will break, but maybe
there needs to be an alternate testing strategy (mocks, etc) for those anyway.

awong-dev commented:
I updated it again, removing the timeout and instead used a Content Security Policy to just kill off external resource loads...cause I like big hammers.

awong-dev commented:
@akainic yup. http://www.w3schools.com/html/html5_webstorage.asp

sessionStorage and localStorage in Javascript just look like a key/value pair on the window.sessionStorage and window.localStorage objects. (There's also some other stuff like .clear() to wipe it out).

Both are origin-scoped storage items meaning the object persists for the origin and only that origin. If you navigate the page away to a different origin, the localStorage and sessionStorage objects will reflect the key/value pairs for that origin, similar to a cookie.  The main difference between localStorage and sessionStorage is lifetime of the keys. LocalStorage lasts kinda forever...like a cookie w/o an expiration.  SessionStorage lasts for a "session" where the definition of a "session" is pretty complicated. At the very base case though, the session isn't supposed to last beyond a browser restart.

awong-dev commented:
Wait... @akainic I misread your question completely. Let me double check.

awong-dev commented:
https://www.w3.org/TR/webstorage/#the-storage-interface

Under section 4.1:  `The supported property names on a Storage object are the keys of each key/value pair currently present in the list associated with the object.`

It's a little vague, but later it talks about the properties being set, etc.  If you'd like, I can switch the code to explicitly use getItem and setItem.

akainic commented:
@awong-dev If it works I think it's fine! Totally up to you

akainic commented on _health-care/_js/_form.js:
```diff
@@ -0,0 +1,348 @@
+    formRoot.addEventListener("submit",
+        function (e) {
+          e.preventDefault();
+          console.log(build1010ezXml(formRoot));
```
Just curious, what's the reason for outputting the xml doc to the console?

awong-dev replied:
Right now, it's just debugging. This code doesn't actually work yet....

awong-dev replied:
Add a TODO.  BTW, I am LOVING the thoroughness of your review here. 

Like seriously <3-ing.

Thank you!!!!!!

akainic replied:
Ah okay

akainic replied:
@awong-dev No problem!

awong-dev commented:
Yeah...that spec language makes me queezy. As I read the language, there's no requirement that the Storage object accept assignment to non-existing key/value pairs as an alternative for calling setItem(). I will change it.

Very nice catch.

awong-dev committed:
Use sessionStorage.setItem() over sessionStorage[]

Per @akainic, the sessionStorage is mostly defined in terms of setItem
and getItem. In fact, 4.1 of
https://www.w3.org/TR/webstorage/#dom-storage-setitem only says:

   The supported property names on a Storage object are the keys of each
   key/value pair currently present in the list associated with the object.

This could be interpreted to mean that the property only exists after
setItem() has been called.

Chrome 47 does indeed make sessionStorage["foo"] = "bar" equivalent to
sessionStorage.setItem("foo", "bar") but why chance it for syntactic
sugar?

awong-dev committed:
Add TODO for debugging logic.

akainic commented:
@awong-dev LGTM! Still getting set up locally so I haven't tested it, but reading through it looks good

awong-dev commented:
Cool.  FYI, you can click on the "deployed" link in the github above to see a automatic deploy of the code to a testing server. That way you don't need to patch down to examine it.

If it looks good, please merge and delete the branch :)

(Also, if getting set up locally for this specific file takes more than a clone plus `rake install` let me know. That means I scripted something wrong.)

akainic commented:
Great! I will test it a bit on the testing server.

akainic commented on _health-care/_js/_form.js:
```diff
@@ -0,0 +1,349 @@
+  };
+
+  var saveForm = function(formRoot, e) {
+    sessionStorage.setItem("voa_form", JSON.stringify($(formRoot).serializeArray()));
```
Just realized we're not actually doing anything with `formRoot` here so it doesn't get updated on `save`. `sessionStorage` does, but not `formRoot`.

awong-dev replied:
At this point, I don't think there is any UI indication of the auto-save so I don't have anything to update.

My thought is saveForm() went from input fields -> sessionStorage. Then Submit when from sessionStorage -> XML (which should immediately be posted to a server, but haven't coded that up).

akainic replied:
Makes sense, so we probably don't need to pass `formRoot` to the function then?

awong-dev replied:
It's a bit of a bias, but I like making my functions as "pure functional" as possible and passing everything through the arguments.  Makes it easier to refactor IMO because all data dependencies are super explicit.

akainic replied:
Ah I see, makes sense!

akainic commented:
@awong-dev Tested it in staging, and mostly I think what's intended to work will work! Obviously still need validations, the AJAX request to submit the validated form data, and some styling help :). The only thing I noticed I commented on above, that the xml isn't updated on form change or submit.

awong-dev commented:
Yay! You wanna do the merge honors?
